opt/memo: no volatility defined between tuple and tuple

[(roachtest).sqlsmith/setup=empty/setting=no-ddl failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2219091&tab=buildLog) on [master@300dbe9a958a9461f320e262b0929cf2c44b7c24](https://github.com/cockroachdb/cockroach/commits/300dbe9a958a9461f320e262b0929cf2c44b7c24):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=no-ddl/run_1
	sqlsmith.go:172,sqlsmith.go:202,test_runner.go:754: error: pq: internal error: no volatility for cast tuple{geometry, oid[], int[]}::tuple{geometry, oid[], int2[]}
		stmt:
		WITH
			with_2139 (col_11421) AS (SELECT * FROM (VALUES (NULL), (ARRAY[]:::OID[])) AS tab_6573 (col_11421)),
			with_2140 (col_11422) AS (SELECT * FROM (VALUES (15963:::INT8)) AS tab_6574 (col_11422))
		SELECT
			tab_6576.col_11426 AS col_11427
		FROM
			(
				VALUES
					(
						('01050000000300000001020000000200000054E3D9DAC83FD3C7A8DB49C14AE9CD471659D1796DD7D8C7803C0B6F5881D34701020000000400000080CDD4FCF098E1473863CBA5BECAB2C7D2A0B00508F7EB4748B318A4F9E9C447D81ED8FA89A6E7C7C8C7A2020AD7C4470836C9A6E747EAC7805324EDA61C79C701020000000B000000F66865EA17E1DBC7CA2BF67951BFDBC73B328C248C77E8C7D2754A32793DEFC7A0EE2D7F61A3C9C732B13428C358EBC700FFB36B958C834750DC3323537CD6C7921549F7511BED47EAE8E0A0A05DDCC7202EA600442ECE47F4C85B7E8C65C4C7F4A365211B71E347000872B9E6416EC7D02C68BFA25DB847C07F47EE857CE9475841E2B58952CAC7D487B468AB83E347F2F9E9993CE8D0C798CF9421972AD0476A2DE8CB17D5E3C7D052FB000229BD47':::GEOMETRY, ARRAY[4070982578:::OID,837520232:::OID,1319489961:::OID,1573271088:::OID,3402151579:::OID], ARRAY[3815:::INT8,(-19545):::INT8,(-10048):::INT8,(-1):::INT8]),
						B'000000000010101111101000000'
					),
					(
						CASE
						WHEN NULL
						THEN (
							SELECT
								('0105000000090000000102000000020000007C95DE97BCD9E44756805560F03BEA4764691133F946E2C7B80706267190E747010200000003000000A85F73C4B045DA47B86674F2BA09D147946819ACDB07E8475CE7626647DDE047A073E92196ADE2477A903AAA9F57E547010200000002000000D40FA8EA144FEFC7100AD51A9182E147131AD061B8DFEFC7907BB7805D6DCD470102000000030000005160DFAAC5B7EFC75586F59BE67BE1C752659EC8B038EFC71068DF418D30EDC7204280528E5EB8C772FDEBE23CFAE6C701020000000200000098178D0609C3E14742B8A3727DE3E0C772D01BA548AAED478E580795B715D2C7010200000002000000580AB1F32C58D6477C44EBFBFE5ED7C708FE7F7C7C4FE3470077CC5644838347010200000003000000381D405C1301BEC7803CA388EFFE95C74CD726E59671D5C742B4B56163C7D2C73085A0A244DBB1C7FCDB0C3589BEE2C70102000000090000004CFBC31096F4DAC77844124AAD7FC547A40FB3DF9A97E2C700FFA705632495C7C0F9EF155F59EBC77C45F81AB8A2E6C72AE988491787D3C778F063357A89D3C7E8877701CF85D8C7328B96987D44E6C7B0DF6DCD41DBCA4764453815FAE7D647BEA8BBA407DBE84750F52BE28F03E747682F7E7E026AD44750250CE1F5B3E94760EBA220D2C9B8470EFCF511D7C6E1470102000000030000008C3B85AC8CD1D847A84C513545EEEC4790106C8C734CB64790C66D22975FE14793838A42017EE3C7BCA413CEB3A5E247':::GEOMETRY, ARRAY[1481278612:::OID,2641891697:::OID,2100605399:::OID], ARRAY[]:::INT2[])
									AS col_11424
							FROM
								(VALUES (NULL), (1737527083:::OID)) AS tab_6575 (col_11423)
							WHERE
								true
							LIMIT
								1:::INT8
						)
						ELSE ('0101000000BC8BC6C28BB7CCC71C70824E7238EAC7':::GEOMETRY, ARRAY[2774504621:::OID,0:::OID], ARRAY[(-23791):::INT8])
						END,
						B'000001100000000110000001101'
					),
					(
						('0106000000020000000103000000010000000600000000E5EAC7FCFE94473B5CABF8DB3CEDC73862CDD26B8DCA4708F531C130CCECC7F48409D6CA93EE47E064DB341920D247E844F096C9FEC2C7EED51139B594E647B1DB3B1AFB2DECC76056FBF48944D64700E5EAC7FCFE94473B5CABF8DB3CEDC701030000000100000005000000C0573A23EE1E984778E5CA042794DCC7FAE90EF7B82DE347181AA2518675CD4750C28252B063DF474864F8F8AA4FE54762ECFF3FA51FE6C7F0988D78F67DE047C0573A23EE1E984778E5CA042794DCC7':::GEOMETRY, ARRAY[0:::OID,1906831471:::OID,4008982506:::OID,2563889166:::OID], ARRAY[28939:::INT8,(-31020):::INT8]),
						B'100001100001110101110111101'
					)
			)
				AS tab_6576 (col_11425, col_11426)
		WHERE
			false
		GROUP BY
			tab_6576.col_11426
		HAVING
			st_coveredby('0104000020E61000000800000001010000005070C4561F3520C048AAA89C04DF21400101000000E83C2B2250B03EC0B4A37576870B504001010000008058E14E771A25C028EC064F6BAD31C0010100000048E258D26D834440807B5640FC724A4001010000005AF378DFF6AD654006EB51B34D5C52400101000000CCDCA97EE56D5D40000827B8CDB7B33F0101000000AB76889192E255C070E41D437E194C40010100000090109BCC4BA820C0EC1FDE61A02E3740':::GEOGRAPHY::GEOGRAPHY, NULL::GEOGRAPHY)::BOOL
		ORDER BY
			tab_6576.col_11426 DESC;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=no-ddl](https://teamcity.cockroachdb.com/viewLog.html?buildId=2219091&tab=artifacts#/sqlsmith/setup=empty/setting=no-ddl)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Dno-ddl.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/array
+++ b/pkg/sql/logictest/testdata/logic_test/array
@@ -1741,3 +1741,12 @@ SELECT x FROM t GROUP BY x
 {1,2}
 {NULL,NULL}
 {1,2,NULL,4,NULL}
+
+# Regression test for #53395: cast between tuples.
+query T
+SELECT CASE
+  WHEN NULL THEN (('1':::INT8, ARRAY[]:::INT2[]))
+  ELSE ('2':::INT8, ARRAY[(-23791):::INT8])
+END
+----
+(2,{-23791})

--- a/pkg/sql/sem/tree/casts.go
+++ b/pkg/sql/sem/tree/casts.go
@@ -328,6 +328,7 @@ func init() {
 
 // lookupCast returns the information for a valid cast.
 // Returns nil if this is not a valid cast.
+// Does not handle array and tuple casts.
 func lookupCast(from, to types.Family) *castInfo {
 	return castsMap[castsMapKey{from: from, to: to}]
 }
@@ -339,6 +340,25 @@ func LookupCastVolatility(from, to *types.T) (_ Volatility, ok bool) {
 	// Special case for casting between arrays.
 	if fromFamily == types.ArrayFamily && toFamily == types.ArrayFamily {
 		return LookupCastVolatility(from.ArrayContents(), to.ArrayContents())
+	}
+	// Special case for casting between tuples.
+	if fromFamily == types.TupleFamily && toFamily == types.TupleFamily {
+		fromTypes := from.TupleContents()
+		toTypes := to.TupleContents()
+		if len(fromTypes) != len(toTypes) {
+			return 0, false
+		}
+		maxVolatility := VolatilityLeakProof
+		for i := range fromTypes {
+			v, ok := LookupCastVolatility(fromTypes[i], toTypes[i])
+			if !ok {
+				return 0, false
+			}
+			if v > maxVolatility {
+				maxVolatility = v
+			}
+		}
+		return maxVolatility, true
 	}
 	cast := lookupCast(fromFamily, toFamily)
 	if cast == nil {

--- a/pkg/sql/sem/tree/casts_test.go
+++ b/pkg/sql/sem/tree/casts_test.go
@@ -152,3 +152,54 @@ func TestCastsFromUnknown(t *testing.T) {
 		}
 	}
 }
+
+func TestTupleCastVolatility(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testCases := []struct {
+		from, to []*types.T
+		exp      string
+	}{
+		{
+			from: nil,
+			to:   nil,
+			exp:  "leak-proof",
+		},
+		{
+			from: nil,
+			to:   []*types.T{types.Int},
+			exp:  "error",
+		},
+		{
+			from: []*types.T{types.Int},
+			to:   []*types.T{types.Int},
+			exp:  "immutable",
+		},
+		{
+			from: []*types.T{types.TimestampTZ},
+			to:   []*types.T{types.Date},
+			exp:  "stable",
+		},
+		{
+			from: []*types.T{types.Int, types.TimestampTZ},
+			to:   []*types.T{types.Int, types.Date},
+			exp:  "stable",
+		},
+	}
+
+	for _, tc := range testCases {
+		from := *types.EmptyTuple
+		from.InternalType.TupleContents = tc.from
+		to := *types.EmptyTuple
+		to.InternalType.TupleContents = tc.to
+		v, ok := LookupCastVolatility(&from, &to)
+		res := "error"
+		if ok {
+			res = v.String()
+		}
+		if res != tc.exp {
+			t.Errorf("from: %s  to: %s  expected: %s  got: %s", &from, &to, tc.exp, res)
+		}
+	}
+}

